cloudinit/net/__init__.py

```diff
@@ -491,24 +545,34 @@ def find_fallback_nic_on_linux(blacklist_drivers=None):
             possibly_connected.append(interface)
             continue
 
-    # don't bother with interfaces that might not be connected if there are
-    # some that definitely are
-    if connected:
-        potential_interfaces = connected
-    else:
-        potential_interfaces = possibly_connected
-
-    # if eth0 exists use it above anything else, otherwise get the interface
-    # that we can read 'first' (using the sorted definition of first).
-    names = list(sorted(potential_interfaces, key=natural_sort_key))
-    if DEFAULT_PRIMARY_INTERFACE in names:
-        names.remove(DEFAULT_PRIMARY_INTERFACE)
-        names.insert(0, DEFAULT_PRIMARY_INTERFACE)
-
-    # pick the first that has a mac-address
-    for name in names:
-        if read_sys_net_safe(name, "address"):
-            return name
+    # Order the NICs:
+    # 1. DEFAULT_PRIMARY_INTERFACE, if connected.
+    # 2. Remaining connected interfaces, naturally sorted.
+    # 3. DEFAULT_PRIMARY_INTERFACE, if possibly connected.
+    # 4. Remaining possibly connected interfaces, naturally sorted.
+    sorted_interfaces = []
+    for interfaces in [connected, possibly_connected]:
+        interfaces = sorted(interfaces, key=natural_sort_key)
+        if DEFAULT_PRIMARY_INTERFACE in connected:
```

Comment:
This should be `interfaces` rather than `connected`, correct?

Reply:
Ah yes, indeed, thanks!